src/routes/actions.js

```diff
@@ -10,39 +13,130 @@ class Actions {
+   * @param {Object} previousFields The fields we want to compare
+   * @see getHookLoadController() for usage
+   */
+  static areFieldsConsistant(fields, previousFields) {
+    const containsAll = (left, right) =>
+      Object.keys(left).every((key) => Object.keys(right).includes(key));
+    return containsAll(fields, previousFields) && containsAll(previousFields, fields);
+  }
+
+  /**
+   * Generate a callback for express that handles the `values` route.
+   * This legacy route is used to handle smart actions' forms initialization.
+   * Users should use the `load` hook now.
+   *
+   * @param {*} action The smart action
+   * @returns {Function} A route callback for express
+   */
+  getValuesController(action) {
     return (request, response) => {
-      let values = action.values ? action.values(request.body.data.attributes.values) : {};
+      const successResponse = (object) => response.status(200).send(object);
+      const values = action.values ? action.values(request.body.data.attributes.values) : {};
 
-      if (!(values.then && typeof func === 'function')) {
-        return response.status(200).send(values);
-      }
+      if (!(values.then && typeof func === 'function')) return successResponse(values);
 
       return values
-        .then((valuesComputed) => {
-          values = valuesComputed;
-        })
+        .then((valuesComputed) => successResponse(valuesComputed))
         .catch((error) => {
           this.logger.error(`Cannot send the values of the "${action.name}" Smart Actions because of an unexpected error: `, error);
-          values = {};
-        })
-        .finally(() => response.status(200).send(values));
+          return successResponse({});
+        });
     };
   }
 
+  /**
+   * Generate a callback for express that handles the `load` hook.
+   *
+   * @param {*} action The smart action
+   * @param {*} model The model of the smart action
+   * @param {*} Implementation Gives access to current Implementation (mongoose or sequelize)
+   * @returns {Function} A route callback for express
+   */
+  getHookLoadController(action, model, Implementation) {
+    return async (request, response) => {
+      const recordId = request.body.data.attributes.recordsId[0];
+      const record = await new Implementation.ResourceGetter(model, { recordId }).perform();
+      const fields = Object.fromEntries(
+        action.fields.map((field) => [field.field, { ...field, value: null }]),
+      );
+
+      try {
+        if (typeof action.hooks.load !== 'function') throw new Error('load must be a function');
+
+        const result = await action.hooks.load({ record, fields });
+
+        if (!(result && typeof result === 'object')) {
+          throw new Error('load hook must return an object');
+        } else if (!Actions.areFieldsConsistant(result, fields)) {
+          throw new Error('fields must be unchanged (no addition nor deletion allowed)');
+        }
```

Comment:
We also need to check no extra properties on fields have been added or some have been removed.

Reply:
`areFieldsConsistant` validate this case.

Reply:
Does `areFieldsConsistant` check if there is no addition/deletion of fields AND no addition/deletion of fields properties?
This is linked to: https://github.com/ForestAdmin/forestadmin/pull/2688/files#diff-0e15c5d2b7cb907c2a110c883d828d030e3cb31c75f66a44858a11a89d13d1ddR49

Reply:
> no addition/deletion of fields properties?

It is not that simple due to various reasons. let's talk about it when you want, then we will make a decision.

Reply:
Resolved